src/NServiceBus.Extensions.Hosting/NServiceBusHostedService.cs

```diff
@@ -4,29 +4,44 @@
     using System.Threading;
     using System.Threading.Tasks;
     using Microsoft.Extensions.Hosting;
-    using NServiceBus;
 
     class NServiceBusHostedService : IHostedService
     {
-        public NServiceBusHostedService(IStartableEndpointWithExternallyManagedContainer startableEndpoint, IServiceProvider serviceProvider)
+        public NServiceBusHostedService(IStartableEndpointWithExternallyManagedContainer startableEndpoint)
         {
             this.startableEndpoint = startableEndpoint;
-            this.serviceProvider = serviceProvider;
         }
 
+        public Task<IEndpointInstance> Endpoint => endpointTcs.Task;
+
         public async Task StartAsync(CancellationToken cancellationToken)
         {
-            endpoint = await startableEndpoint.Start(new ServiceProviderAdapter(serviceProvider))
-                .ConfigureAwait(false);
+            try
+            {
+                var endpoint = await startableEndpoint.Start(new ServiceProviderAdapter(serviceProvider))
+                    .ConfigureAwait(false);
+                endpointTcs.SetResult(endpoint);
```

Comment:
Is it a problem when the continuation of the one that awaits the TCS completes on the thread that enters `StartAsync`? If yes then you need to declare `TaskContinuationOptions.RunAsynchronously` or wrap `SetXYZ` with a `Task.Run`

Reply:
I think there should be no issue as asp.net core doesn't use a sync context anymore?

Reply:
That's true but the host could that runs this code right?

I'm fine of leaving as is if you think it is a non issue

Reply:
which host are you referring to? I've also considered to use `ConfigureAwait(true)` in most paths to ensure that we're not breaking something and personally I'm perfectly fine with that (safer?) approach.

Reply:
I think `ConfigureAwait(false)` is a good default because we are library code and it only affects our continuations.

With the TCS I'd go with the safer approach here especially because we only pay the overhead of Task.Run once